python/collapse.py

```diff
@@ -0,0 +1,197 @@
+import mgp
+
+from collections import deque
+from typing import Dict, List, Tuple, Set, Optional
+
+
+@mgp.read_proc
+def collapse(
+    context: mgp.ProcCtx,
+    vertices: mgp.List[mgp.Vertex],
+    collapse_edge_types: mgp.List[str],
+    collapse_pseudo_node_labels: mgp.Nullable[List[str]] = None,
+) -> mgp.Record(from_vertex=mgp.Vertex, path=mgp.Path, to_vertex=mgp.Vertex):
+    """
+    Returns collapsed graph.
+
+    Returned collapsed graph is a list of records: (from_node, path, to_node).
+    Algorithm collapses all related vertices into a one of the top vertices.
+    List of top vertices is given by vertices parameter
+    If edge has one of the types defined in collapse_edge_types parameter,
+    edge to_vertex will be collapsed in the same top vertex as edge from_vertex.
+
+    collapse_pseudo_node_labels is optional parameter. This parameter must be used
+    if top vertices are connected with pseudo vertex
+    (vertices with exactly one input edge and exactly one output edge).
+
+    Path can contain one edge or two edges.
+    Path contains two edges in a pseudo vertex case.
+
+    Procedure will raise error if any of pseudo vertices defined by collapse_pseudo_node_labels
+    parameter will have more or less than one input and one output edges.
+
+    Procedure will raise an error if collapsed groups aren't disjunctive sets.
+
+    Example call:
+        MATCH (n:Plant) WITH COLLECT(n) AS nodes
+        CALL collapse.collapse(nodes, ["OWN"], ["Transport"])
+        YIELD from_vertex, path, to_vertex
+        RETURN from_vertex, nodes(path), to_vertex;
+    """
+    union_find = dict()
+    node_to_edge = (
+        set(collapse_pseudo_node_labels) if collapse_pseudo_node_labels else None
+    )
+    vertices_id = {v.id for v in vertices}
+    vertex_neighbours = []
+    visited_global = set()
+    for v in vertices:
+        paths, group = _bfs(
+            v, vertices_id, collapse_edge_types, visited_global, node_to_edge
+        )
+        visited_global.update([child_id for child_id, _ in group.items()])
```

Comment:
If you are only using keys from dict, then call `group.keys()` without ignoring the other parameter